### PR TITLE
Add validation to axis for tf.nn.softmax

### LINKED ISSUE #22793: tf.nn.softmax can give a result, when input shape is [2,3] and axis=2, Is it a bug?
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
windows 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
1.10.0
- **Python version**:
3.5.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
no CUDA
- **GPU model and memory**:
no GPU
- **Exact command to reproduce**:

### Describe the problem

for  `tf.nn.softmax` , the `axis=2` but the input tensor shape is `[2,3]`.
And the result is 

```
tf.Tensor(
[[0.5 0.5 0.5]
 [0.5 0.5 0.5]], shape=(2, 3), dtype=float32)
```

I think the axis should `>=0` and `<=1`

if the `axis>=2`, it should be error.

### Source code / logs

```
import tensorflow as tf
tf.enable_eager_execution()
ones = tf.ones(shape=[2,3])
temp = tf.nn.softmax(ones,axis=2)
print(temp)
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #22793 where
an invalid axis (outside of `[-dim, dim)`) still returns
value. This behavior is different from most other ops in
tf like `tf.argmax`/etc.

This fix add the validation of axis so that an error
will be returned in case of invalid axis.

This fix fixes #22793.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/ops/nn_ops.py [hunk 1]
```diff
@@ -1679,22 +1680,25 @@ def _swap_axis(logits, dim_index, last_index, name=None):
 
   # If dim is not the last dimension, we have to do a transpose so that we can
```
Comment (hawkinsp) [inline tensorflow/python/ops/nn_ops.py]: Thanks for the PR! But can't we do something even more direct here?

This code already assumes that shape.ndims is not None, and that dim is a Python integer. So I don't think we need to do this test as part of the graph — we can do it in Python instead, which is simpler and easier to understand. i.e., I think you can just write:

```python
if dim < -shape.ndims or dim >= shape.ndims:
  raise ... etc.
```

This would also give the caller a more usable backtrace from Python, rather than a somewhat more cryptic error when running the graph.

**Comment by yongtang** [comment]
Thanks @hawkinsp for the review. The PR has been updated, now validation does not go through graph any more. Please take a look and let me know if there are any issues.

**Review by hawkinsp — APPROVED** [review]
Looks good to me, assuming the tests pass. Thanks for the PR!

**Comment by yongtang** [comment]
@harshini-gadige It looks like all tests passed. Maybe `ready to pull` could be added?
